Add structures for the enrollment process

linked0 opened the pull request:
- Enrollment: Defines data for enrollment process
- EnrollmentManager: Handles data and process for enrollment
- unittest code

#406

TrustHenry commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,138 @@
+import agora.common.crypto.Schnorr : Pair;
+
+import std.random;
+import std.stdio;
```
`import std.stdio;`
Not used here.

TrustHenry commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,138 @@
+    auto enroll_data = man.getEnrollmentData();
+    auto enroll_data2 = man.getEnrollmentData();
+    assert(enroll_data.S != enroll_data2.S);
+}
```
An Enter required in the last line.

AndrejMitrovic commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,138 @@
+        this.data.K = this.node_keypair.address;
+
+        // cycle length
+        this.data.n = uniform(0, 1008, this.genint);
```
The number of voting rounds the node participates in should not be random. It should be set by the node operator.

Currently the Yellow paper says:

> 3. Enrollment process
> n (cycle length): the number of rounds a validator will participate in (currently fixed to (freezing period / 2)];

For now, I suggest making N a constructor parameter for `EnrollmentManager`.

linked0 replied:
Thank you for your comment.
I'll change the number as described in the yellow paper, but I'd like to set the number inside the class not by the parameter of the constrcutor because the setting the number can be a burden to other classes such as the Node class. By a burden, I mean determining which number should be set. 
Any opinion about this would be helpful.

AndrejMitrovic replied:
Ok, you can keep this internal for now.

AndrejMitrovic commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,138 @@
+
+    private static void round (ref Hash image, size_t n)
+    {
+        image = hashFull(image);
```
This can be replaced with a foreach loop.

```D
foreach (i; 0 .. n)
    image = hashFull(image);
```

AndrejMitrovic commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,138 @@
+
+    /// n: cycle length, A number within bounds [0; 1008]
+    /// original spec: [0; (Balance(K) % 40,000) * freezing period * 2]
+    public ulong n;
```
This can be `uint`

AndrejMitrovic commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,138 @@
+        this.data.n = uniform(0, 1008, this.genint);
+
+        // random seed
+        this.data.X = hashFull(this.node_keypair.secret.data);
```
This is the part that should use the random number generator.

AndrejMitrovic replied:
The random seed should not be based on any existing data, it should ideally just be a random value. For example:

```D
Hash X = Scalar.random().hashFull();
```

The initial X should be stored somewhere locally so the Nth preimage can be generated from it on demand when a node needs to reveal that Nth preimage later. Either that, or you can just store an array of preimages for quick lookup later.

AndrejMitrovic replied:
any update on this?

linked0 replied:
Ah, Mathias commented that I could use the constant value, or 1008 at the moment before he left.

AndrejMitrovic replied:
I'm talking about the random seed value X, not N. We should use a random number generator, and not a hash of the private data.

AndrejMitrovic replied:
Just to update everyone on this: Initially we planned on using the private key as the seed, which would have to be revealed at the end of the validation cycle. However, after @Geod24 had reservations (even though this part is still in the yellowpaper), so for now we can just make this a random number.

The reason why we need this X, is for the verifiable random number generation in the blockchain. This is important for many things: quorum balancing (https://github.com/bpfkorea/agora/issues/240), contracts (think games, lottery apps), etc.

AndrejMitrovic commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,138 @@
+    {
+        this.node_keypair = keypair;
+        this.balance = balance;
+        this.genint = Random(123);
```
This is not a safe way to use a random-number generator. The `123` part is the seed. The sequence of numbers you get out of a random-number generator is always the same for the given seed.

Instead, I suggest using `randombytes_uniform()` from libsodium to generate a number within a given range.

AndrejMitrovic commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,138 @@
+
+        // signature
+        Hash hash = hashFull(*this.data);
+        this.data.S = this.node_keypair.secret.sign(hash[]);
```
The signature part actually has to be a schnorr signature using `r` (the `random_key.v` scalar) above. If the enrollment is not signed with `r`, then you can't prove that you actually know `R`. :)

https://github.com/bpfkorea/agora/blob/5ef14fe0ce879da52cc47716d7688bf5ab0d943f/source/agora/common/crypto/Schnorr.d#L196

^ This function can be used. It's currently private, but I'll make it public in a PR in a moment.

linked0 commented:
Ready for review.

MichaelKim20 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,130 @@
+    Defines data for enrollment process
+*******************************************************************************/
+
+public struct EnrollmentData{
```
```suggestion
public struct EnrollmentData
{
```

MichaelKim20 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,130 @@
+            keypair = the keypair of the owner node
+            balance = account balance of the owner node
+    ***************************************************************************/
+    public this(KeyPair keypair, Amount balance)
```
```suggestion
    public this (KeyPair keypair, Amount balance)
```

MichaelKim20 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,130 @@
+        Returns:
+            a EnrollmentData object
+    ***************************************************************************/
+    public EnrollmentData createEnrollmentData()
```
```suggestion
    public EnrollmentData createEnrollmentData ()
```

MichaelKim20 commented on source/agora/common/crypto/ECC.d:
```diff
@@ -308,6 +308,19 @@ public struct Point
+            dg = Hashing function accumulator
+
+    ***************************************************************************/
+    public void computeHash(scope HashDg dg) const nothrow @safe @nogc
```
```suggestion
    public void computeHash (scope HashDg dg) const nothrow @safe @nogc
```

MichaelKim20 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,130 @@
+import agora.common.Types;
+import agora.common.crypto.Schnorr : Pair;
+import agora.common.crypto.Schnorr : Signature;
+import agora.common.crypto.Schnorr : sign;
```
**According to `Agora coding style`**
imports are organized by packages: agora first, dependencies second, std third, core last. Within packages, alphabetical order is prefered, with modules taking preference over packages, e.g.:

MichaelKim20 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,130 @@
+        Makes a data object for enrollment process
+        Returns:
+            a EnrollmentData object
+    ***************************************************************************/
```
```
    /***************************************************************************
    
        Makes a data object for enrollment process
        
        Returns:
            a EnrollmentData object
            
    ***************************************************************************/

    public EnrollmentData createEnrollmentData ()
```

MichaelKim20 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,130 @@
+        Params:
+            keypair = the keypair of the owner node
+            balance = account balance of the owner node
+    ***************************************************************************/
```
```
    /***************************************************************************
    
        Constructor
        
        Params:
            keypair = the keypair of the owner node
            balance = account balance of the owner node
            
    ***************************************************************************/
    
    public this (KeyPair keypair, Amount balance)
```

MichaelKim20 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,130 @@
+        All rights reserved.
+    License:
+        MIT License. See LICENSE for details.
+*******************************************************************************/
```
```
/*******************************************************************************

    Contains supporting code for tracking the current ledger.

    Copyright:
        Copyright (c) 2019 BOS Platform Foundation Korea
        All rights reserved.

    License:
        MIT License. See LICENSE for details.

*******************************************************************************/
```

MichaelKim20 reviewed (changes requested):
Well done. I'd like to make some suggestions.
I reviewed it according to `Agora coding style`

AndrejMitrovic commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,426 @@
+
+        // signature
+        Hash hash = hashFull(*this.data);
+        this.data.S = sign(this.schnorr_pair, hash[]);
```
I don't think this is right. Here you are proving ownership of `R` by using `r` as a signing key, and randomly generating the actual R in the signature (see the `sign` method overload used here, it randomly generates R). But, this doesn't prove that you know `k` of `K`.

Instead I suggest signing it like so:

```D
auto pk = Scalar(this.node_keypair.seed[]);
this.data.S = sign(kp, kp.toPoint(), this.schnorr_pair.V, this.schnorr_pair.v, hash[]);
```

linked0 commented:
Ready for review.

MichaelKim20 commented on source/agora/consensus/data/EnrollmentData.d:
```diff
@@ -0,0 +1,111 @@
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
```
```
/*******************************************************************************

    Contains supporting code for a enrollment data used in enrollment process.

    Copyright:
        Copyright (c) 2019 BOS Platform Foundation Korea
        All rights reserved.

    License:
        MIT License. See LICENSE for details.

*******************************************************************************/
```

MichaelKim20 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,426 @@
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
```
```
/*******************************************************************************

    Contains supporting code for enrollment process.

    Copyright:
        Copyright (c) 2019 BOS Platform Foundation Korea
        All rights reserved.

    License:
        MIT License. See LICENSE for details.

*******************************************************************************/
```

MichaelKim20 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,426 @@
+                        :memory: was passed
+            keypair = the keypair of the owner node
+
+    ***************************************************************************/
```
```
  /***************************************************************************

        Constructor

        Params:
            db_path = path to the database file, or in-memory storage if
                        :memory: was passed
            keypair = the keypair of the owner node

    ***************************************************************************/
```

MichaelKim20 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,426 @@
+
+    public bool acceptEnrollment(EnrollmentData enroll) @safe
+    {
+	    if (!this.add(enroll))
```
```suggestion
        if (!this.add(enroll))
```

MichaelKim20 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,426 @@
+
+    ***************************************************************************/
+
+    public bool acceptEnrollment(EnrollmentData enroll) @safe
```
```suggestion
    public bool acceptEnrollment (EnrollmentData enroll) @safe
```

MichaelKim20 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,426 @@
+
+    public bool hasEnrollmentDataHash (EnrollmentData enroll) @trusted
+    {
+	    auto enroll_hash = hashFull(enroll.K); 
```
```suggestion
        auto enroll_hash = hashFull(enroll.K); 
```

MichaelKim20 reviewed (changes requested):
I'd like to make some suggestions.

linked0 commented:
Ready for review. Any suggestions are welcome.

linked0 commented:
Ready for review.
I have already implemented the API endpoint on the Node class in preparation for next PR. The next PR will be opened in next Monday. I can apply other suggestions to those classes in the next PR.

AndrejMitrovic commented on source/agora/consensus/data/EnrollmentData.d:
```diff
@@ -0,0 +1,113 @@
+    Defines data for enrollment process
+*******************************************************************************/
+
+public struct EnrollmentData{
```
```suggestion
public struct EnrollmentData
{
```

AndrejMitrovic commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,469 @@
+    {
+        static import core.atomic;
+        static shared bool initialized;
+        if (core.atomic.cas(&initialized, false, true))
```
The SQLite library is already initialized in another place. It shouldn't be initialized twice. I suggest removing this code.

AndrejMitrovic commented on source/agora/common/crypto/ECC.d:
```diff
@@ -178,6 +178,20 @@ public struct Scalar
+
+    ***************************************************************************/
+
+    public void computeHash (scope HashDg dg) const nothrow @safe @nogc
```
This should be in a separate commit.

AndrejMitrovic commented on source/agora/consensus/data/EnrollmentData.d:
```diff
@@ -0,0 +1,113 @@
+    /// R: signature noise, The initial nonce used for signing
+    public Point R;
+
+    /// S: A signature for the messge H(K, X, n, R) and the key K, using R
```
```suggestion
    /// S: A signature for the message H(K, X, n, R) and the key K, using R
```

AndrejMitrovic commented on source/agora/consensus/data/EnrollmentData.d:
```diff
@@ -0,0 +1,113 @@
+        deserializePart(this.S, dg);
+    }
+
+    /// Support for sorting transactions
```
Why do you need to support sorting?

AndrejMitrovic commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,469 @@
+/// Logger instance
+private Logger log;
+
+/*******************************************************************************
```
```suggestion
/*******************************************************************************

```

AndrejMitrovic commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,469 @@
+private Logger log;
+
+/*******************************************************************************
+    Handles data and process for enrollment
```
```suggestion
    Handles data and process for enrollment

```

AndrejMitrovic commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,469 @@
+        // set information to member variables
+        this.db_path = db_path;
+        this.node_keypair = keypair;
+        this.data = null;
```
```suggestion
```

Unnecessary, pointers are null-initialized.

AndrejMitrovic commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,469 @@
+        if (db_exists)
+            log.info("Loading database from: {}", db_path);
+
+        // note: can fail. we may want to just recover txes from the network instead.
```
This was copy-pasted from existing code and does not apply here.

AndrejMitrovic commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,469 @@
+
+    ***************************************************************************/
+
+    private static extern(C) void loggerCallback (void *arg, int code,
```
This should be removed, the logger for SQLite is set once in another module.

AndrejMitrovic commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,469 @@
+
+    /***************************************************************************
+
+        Walk over the enrollmane data in the set and call the provided delegate
```
```suggestion
        Walk over the enrollment data in the set and call the provided delegate
```

linked0 commented:
Ready for review.

AndrejMitrovic commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,416 @@
+
+    ***************************************************************************/
+
+    public bool add (EnrollmentData enroll_data) @safe
```
This function unconditionally returns `true`. Maybe you could check `hasEnrollmentDataHash` and return `false` early if `hasEnrollmentDataHash` return true?

AndrejMitrovic commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,416 @@
+
+    ***************************************************************************/
+
+    public bool hasEnrollmentDataHash (const ref EnrollmentData enroll) @trusted
```
This should be renamed to `hasEnrollmentData`. From the name it seemed like the user would provide the hash to the function, not the data.

AndrejMitrovic commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,416 @@
+
+    ***************************************************************************/
+
+    public bool acceptEnrollment (const ref EnrollmentData enroll) @safe
```
Now there are two APIs: `add`, and `acceptEnrollment`. Which one should the user use?

These two functions should be merged together.

AndrejMitrovic commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,416 @@
+
+    ***************************************************************************/
+
+    version (unittest) public EnrollmentData[] take (size_t count) @safe
```
This can be `private`, as `private` is still usable from the same module.

AndrejMitrovic commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,416 @@
+
+    ***************************************************************************/
+
+    public this (in string db_path, in string seed_file_path, KeyPair keypair)
```
Missing docs for `seed_file_path`

AndrejMitrovic commented on source/agora/consensus/data/EnrollmentData.d:
```diff
@@ -0,0 +1,102 @@
+
+*******************************************************************************/
+
+public struct EnrollmentData {
```
```suggestion
public struct EnrollmentData
{
```

AndrejMitrovic commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,416 @@
+
+    /***************************************************************************
+
+        Check if a enrollment data hash exists in the transaction pool.
```
```suggestion
        Check if the enrollment data exists in the transaction pool.
```

linked0 commented:
Ready for review.

AndrejMitrovic commented:
Looks good. I'll let @Geod24 review it when he's back.

TrustHenry commented on source/agora/consensus/data/EnrollmentData.d:
```diff
@@ -0,0 +1,99 @@
+public struct EnrollmentData
+{
+    /// K: UTXO key, A public key matching a frozen UTXO
+    public PublicKey K;
```
It is correct to set the frozen UTXO hash here.
It is the only way to tell if you have frozen UTXO.

AndrejMitrovic replied:
Are you asking to change something here?

linked0 replied:
@TrustHenry You mean that a enrollment data must have a information like the UTXO hash by which the frozen UTXO validator(you are now implementing) could find a matching UTXO?

TrustHenry replied:
@linked0 @AndrejMitrovic  
Yes, that’s right.
In the Enrollment process, Publickey cannot be a key.
A Publickey can contain many UTXOs.
We have to choose one frozen UTXO.
This can be a key in Enrollment.

AndrejMitrovic replied:
> We have to choose one frozen UTXO.

True. We could also sum the frozen UTXOs for the given key, but this might complicate the `N` value which is currently hardcoded by definition.

linked0 replied:
Can I apply this request to my next PR? I have already implemented a API endpoint of Node about the enrollment process and I think we need more time to conclude how it can be implemented. So it would be better to merge the code of this PR to the Upstream and to apply Henry's request to next PR.

AndrejMitrovic replied:
Where is the rush? :) If we know the code will be changed after merging we might as well change it before merging. It's very easy to go from the utxo hash to the PublicKey. It's just `utxo_set[utxo_hash].address`.

linked0 replied:
@Geod24 Which do you think is best way to apply @TrustHenry's request between replacing a public key with a UTXO hash or just adding a UTXO hash member in EnrollmentData structure?

TrustHenry commented:
If UTXO hash becomes PREMARY KEY,
We need to implement a way to find the UTXO hash.

TrustHenry commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,394 @@
+unittest
+{
+    KeyPair kp = KeyPair.random();
+    auto random_seed_path = buildPath(getcwd, "random_set.dat");
```
FileIO is not recommended for unittest.

TrustHenry commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,394 @@
+unittest
+{
+    KeyPair kp = KeyPair.random();
+    auto random_seed_path = buildPath(getcwd, "random_set.dat");
```
FileIO is not recommended for unittest.

linked0 commented:
We determined that the enrollment process will use a frozeon UTXO hash instead of the public key of a node as means to identify a validator, which was output at the whiteboarding session.

linked0 commented:
Ready for review.

Geod24 commented:
Can you rebase on v0.x.x ?

linked0 commented:
I have rebased on v0.x.x and I'm ready for review.

Geod24 commented on source/agora/consensus/data/EnrollmentData.d:
```diff
@@ -0,0 +1,99 @@
+public struct EnrollmentData
+{
+    /// K: UTXO hash, A hash of a frozen UTXO
+    public Hash K;
```
I used `K` in the paper for conciseness, but I think we should have more explicit names. This applies to everything in this struct, except `R` maybe, as it's the name used in the Schnorr formula.

Geod24 commented on source/agora/consensus/data/EnrollmentData.d:
```diff
@@ -0,0 +1,99 @@
+
+    ***************************************************************************/
+
+    public void computeHash (scope HashDg dg) const nothrow @safe @nogc
```
You can remove this

Geod24 commented on source/agora/consensus/data/EnrollmentData.d:
```diff
@@ -0,0 +1,99 @@
+
+    ***************************************************************************/
+
+    public void serialize (scope SerializeDg dg) const @safe
```
Remove

Geod24 commented on source/agora/consensus/data/EnrollmentData.d:
```diff
@@ -0,0 +1,99 @@
+
+    ***************************************************************************/
+
+    public void deserialize (scope DeserializeDg dg) @safe
```
Remove

Geod24 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,427 @@
+
+*******************************************************************************/
+
+static this ()
```
Why not use the `mixin` ?

Geod24 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,427 @@
+import agora.common.crypto.Key;
+import agora.common.crypto.Schnorr : Pair;
+import agora.common.crypto.Schnorr : Signature;
+import agora.common.crypto.Schnorr : sign;
```
In D you can do the following:
```D
import agora.common.crypto.Schnorr : Pair, Signature, sign;
```

If you want to keep the selective import.

Geod24 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,427 @@
+    {
+        const db_exists = db_path.exists;
+        if (db_exists)
+            log.info("Loading database from: {}", db_path);
```
I would make this a `trace` call, and I think it'll be better to check for a valid file (e.g. one with data in it) rather than the file existing. It also works better with unittests.

Geod24 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,427 @@
+
+        Params:
+            db_path = path to the database file, or in-memory storage if
+                        :memory: was passed
```
This is a detail of the database. Maybe just say "path to the database file (sqlite3 format)"

Geod24 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,427 @@
+        Params:
+            db_path = path to the database file, or in-memory storage if
+                        :memory: was passed
+            seed_file_path = path to the data file storing randon seed
```
Wrong name. Also, rando*m*

Geod24 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,427 @@
+            "(key BLOB PRIMARY KEY, val BLOB NOT NULL)");
+
+        // load random seed
+        this.random_seed_path = buildPath(data_dir, "random_set.dat");
```
It does IO (disk) access, so it's not suitable for an unittest.
The other part correctly abstract things using `db_path`, but this does not.

Geod24 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,427 @@
+
+    ***************************************************************************/
+
+    public bool add (EnrollmentData enroll_data) @safe
```
```suggestion
    public bool add (const ref EnrollmentData enroll_data) @safe
```

Also, what @AndrejMitrovic said

Geod24 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,427 @@
+
+        () @trusted { assumeSafeAppend(buffer); } ();
+
+        scope SerializeDg dg = (scope const(ubyte[]) data) nothrow @safe
```
Note to myself: We should make this more idiomatic

Geod24 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,427 @@
+
+    ***************************************************************************/
+
+    public size_t length () @safe
```
I'd say `computeLength` to make it obvious that a call isn't cheap.

Geod24 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,427 @@
+
+    ***************************************************************************/
+
+    public int opApply (scope int delegate(Hash hash, EnrollmentData enroll) dg)
```
```suggestion
    public int opApply (scope int delegate(const ref Hash hash, const ref EnrollmentData enroll) dg)
```

Geod24 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,427 @@
+
+    public void remove (const ref EnrollmentData enroll) @trusted
+    {
+        auto hash = hashFull(enroll.K);
```
During the `INSERT`, `K` is inserted directly, but here it is hashed. So I don't think it works, and I think you need an extra unittest :)
Also, the API should probably be based on `K`, not the full enrollment data.

Geod24 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,427 @@
+
+        // R, signature noise
+        this.schnorr_pair = Pair.random();
+        this.data.R = this.schnorr_pair.V;
```
This also needs to be saved

Geod24 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,427 @@
+    private static Signature signHash (const ref Hash frozen_utxo_hash, 
+        const ref Pair schnorr_pair, const ref Hash hash)
+    {
+        auto pk = Scalar(frozen_utxo_hash);
```
Using the `frozen_utxo_hash` should strike anyone as a pretty bad idea

Geod24 commented on source/agora/node/EnrollmentManager.d:
```diff
@@ -0,0 +1,427 @@
+        {
+            enroll_data = deserializeFull!EnrollmentData(row.peek!(ubyte[])(1));
+
+            // break
```
```suggestion
```

Geod24 reviewed (changes requested):
Added a few comments:
- The IO needs to be abstracted away
- The UTXO hash cannot be used as private key
- `r` needs to be saved

linked0 commented:
Ready for review

TrustHenry reviewed (approved):
LGTM

AndrejMitrovic commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,409 @@
+    man.removeEnrollment(utxo_hash);
+    assert(man.getEnrollmentLength() == 1);
+
+    man.shutdown();
```
This calls the shutdown a second time after `scope (exit) utxo_set.shutdown();` already did it. So you can just remove `man.shutdown()` here.

linked0 replied:
The `man.shutdown()` call must be called. So I will add the `scope(exit) man.shutdown() ` below the `auto man = new UnitEnrollMan()` code

AndrejMitrovic commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,409 @@
+
+        // R, signature noise
+        this.signature_noise = Pair.random();
+        this.data.R = this.signature_noise.V;
```
@Geod24 I think I see an optimization opportunity here. We are storing `R` in both the `EnrollmentData` and the `(R, s)` signature pair. Could we omit R from the `EnrollmentData`? After all, `R` is already hashed in the challenge.

linked0 commented:
Ready for review.
I've removed the signature noise variable from the Enrollment structure because it is also stored in a Signature object.

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,407 @@
+import agora.common.crypto.Key;
+import agora.common.crypto.Schnorr : Pair;
+import agora.common.crypto.Schnorr : Signature;
+import agora.common.crypto.Schnorr : sign;
```
Merge those together (actually I'd say, just loose the selective import and make it a regular import)

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,407 @@
+import d2sqlite3.results;
+import d2sqlite3.sqlite3;
+
+import ocean.util.log.Logger;
```
Remove (it is not used AFAICS)

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,407 @@
+
+/*******************************************************************************
+
+    Handle enrollment data and manage the validator set(db)
```
```suggestion
    Handle enrollment data and manage the validators set
```

I'm not sure what the `db` is about. Looks like an implementation detail.

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,407 @@
+
+        // load random seed
+        this.node_enroll_data_path = buildPath(data_dir, "node_enroll_data.dat");
+        loadNodeEnrollmentData();
```
My comment about mixing abstraction still stands. This is not unittestable. Why not make it part of the database ?

linked0 replied:
The table, or the validator_set is for storing other validators' information, so it does not fits to the information which is just the signature noise of the enrollment data for a node itself. Do you mean another table for storing just one row for the signature noise? I'd like to make a memory storage in the EnrollmentManager in case of an unittest. What do you think about this?

Geod24 replied:
I meant another table. There are other options, but this one seems the most straightforward.

linked0 replied:
I think another table is simple and straightforward also. I will make another table named "node_enroll_data".

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,407 @@
+    {
+        const db_exists = db_path.exists;
+        if (db_exists)
+            log.info("Loading database from: {}", db_path);
```
```suggestion
            log.trace("Loading enrollment database from: {}", db_path); 0 
```

Geod24 replied:
Actually I'd say just remove those 3 lines ?

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,407 @@
+        loadNodeEnrollmentData();
+
+        // set information to member variables
+        this.db_path = db_path;
```
This member variable is never used

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,407 @@
+        // check if already exists
+        if (hasEnrollment(enroll.utxo_key))
+        {
+            log.info("Rejected enroll: {}", enroll);
```
The message is too broad. Imagine trying to track down a bug in 3 months and seeing this. You at least want to know *why* the enrollment was rejected. Something like "Rejected already existing enrollment: {}" gives much more information at a glance.

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,407 @@
+
+    public Enrollment createEnrollment (const ref Hash frozen_utxo_hash) @trusted
+    {
+        this.data = new Enrollment();
```
Why is it a pointer ?

Geod24 replied:
Also is this function supposed to be called multiple times ? In what order ? For what purpose ?
The documentation is very generic. The fact that it is stored as a member of this class tells me that either:
- There can only be one
OR
- We cache the last one

If there can only be one, it feels like it shouldn't be called multiple times. If we cache the last created Enrollment data, why do we do that ? Neither the documentation of this function nor the variable says anything.

linked0 replied:
@Geod24 I'd like to discuss with you about this. I wanna split the EnrollmentManager into the ValidatorSet which manage other validators' information and the EnrollmentManager which manage the enrollment data for a node itself.

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,407 @@
+    }
+
+    // sign enrollment data with schnorr key pairs
+    private static Signature signHash (const ref SecretKey secret_key, 
```
This function is clearly not carrying its weight and should be removed, unless I am missing something ?

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,407 @@
+        this.signature_noise = Pair.random();
+
+        // signature
+        Hash hash = hashFull(*this.data);
```
Why the double hashing ?

AndrejMitrovic replied:
What Mathias is saying, is that the `sign` function already hashes the data structure. So it's not necessary to pre-hash it here, and you could even get rid of `signHash` since it's only used in one place.

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,407 @@
+
+    public bool hasEnrollment (const ref Hash enroll_hash) @trusted
+    {
+        auto results = this.db.execute("SELECT key, val FROM validator_set " ~
```
There is an `EXISTS` statement in SQL to allow for short-circuiting, otherwise the query planner execute over the whole table even if it can find a record:
```sql
SELECT EXISTS(SELECT 1 FROM validator_set WHERE key = ?)
```

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,407 @@
+
+    ***************************************************************************/
+
+    public int opApply (scope int delegate(const ref Hash hash, 
```
This function is not tested

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,407 @@
+
+    ***************************************************************************/
+
+    public bool getEnrollment (const ref Hash enroll_hash,
```
This function is not tested

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,407 @@
+        buffer.length = 0;
+
+        // check if already exists
+        if (hasEnrollment(enroll.utxo_key))
```
This branch is not tested

Geod24 commented on source/agora/common/crypto/Schnorr.d:
```diff
@@ -133,6 +135,36 @@ public struct Pair
+
+    ***************************************************************************/
+
+    public void serialize (scope SerializeDg dg) const @safe
```
Superfluous, no ?

AndrejMitrovic commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,407 @@
+
+    ***************************************************************************/
+
+    public bool addEnrollment (Enrollment enroll) @safe
```
I think this method needs to be `nothrow`. Otherwise, you have to care about two cases: if the function returns false, or if it throws an exception. If it returns a status code, we need to make sure it never throws.

Geod24 replied:
Ping

linked0 replied:
I wrote the code with reference to the add function of the TransactionPool. I have tried to add "nothrow" to the function but it output too many errors which might take me too much time. So, I'd like to modify that as Drey suggested after fully understanding the concept in other PR if possible.

linked0 committed:
Add structures for the enrollment process
- Enrollment: Defines data for enrollment process
- EnrollmentManager: Handles data and process for enrollment
- unittest code

linked0 commented:
Ready for review

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,353 @@
+import agora.common.crypto.ECC;
+import agora.common.crypto.Key;
+import agora.common.crypto.Schnorr;
+import agora.common.crypto.Schnorr : Signature;
```
Redundant

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,353 @@
+
+        foreach (row; results)
+        {
+            signature_noise = deserializeFull!Pair(row.peek!(ubyte[])(1));
```
This branch is not tested

linked0 replied:
In the unittest code, all the data saved in the database cannot be shared because the database is in memory as far as I know. So this branch cannot be tested in the unittest code. Would you please show me a clue for that?

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,353 @@
+        buffer.length = 0;
+
+        // check if already exists
+        if (hasEnrollment(enroll.utxo_key))
```
```suggestion
        if (this.hasEnrollment(enroll.utxo_key))
```

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,353 @@
+        serializePart(enroll, dg);
+
+        () @trusted {
+            db.execute("INSERT INTO validator_set (key, val) VALUES (?, ?)",
```
```suggestion
            this.db.execute("INSERT INTO validator_set (key, val) VALUES (?, ?)", 2 
```

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,353 @@
+    public size_t getEnrollmentLength () @safe
+    {
+        return () @trusted {
+            return db.execute("SELECT count(*) FROM validator_set").oneValue!size_t;
```
```suggestion
            return this.db.execute("SELECT count(*) FROM validator_set").oneValue!size_t; 3 
```

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,353 @@
+
+    ***************************************************************************/
+
+    public Enrollment createEnrollment (const ref Hash frozen_utxo_hash) @trusted
```
`const ref` didn't work ?

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,353 @@
+        auto results = this.db.execute("SELECT EXISTS(SELECT 1 FROM validator_set " ~
+            "WHERE key = ?)", enroll_hash[]);
+
+        return results.oneValue!(bool);
```
No need for the intermediate variable, return it directly (it might also be more optimal thanks to NRVO)

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,353 @@
+    private Database db;
+
+    /// Node's key pair
+    private KeyPair node_key_pair;
```
```suggestion
    private Pair node_key_pair;
```

You could also loose the `node_` part if you want, as I don't think there's any other kind of key_pair the node will know of :)

linked0 replied:
The node_key_pair must be KeyPair type because it is set from a node's keypair. This variable is set in the constructor, which will be called by a node later.

AndrejMitrovic replied:
He means to rename `node_key_pair` to `key_pair`

linked0 replied:
His suggested change is about the type of the node_key_pair variable. I have already renamed the name of the vairable to `key_pair` in my current code.

AndrejMitrovic replied:
You're right, I read it wrong.

Geod24 replied:
> The node_key_pair must be KeyPair type because it is set from a node's keypair.

But at the signature steps you turn it into a `Pair`.

Geod24 commented on source/agora/consensus/data/Enrollment.d:
```diff
@@ -0,0 +1,43 @@
+
+import agora.common.crypto.ECC;
+import agora.common.crypto.Key;
+import agora.common.Deserializer;
```
```suggestion
```

Geod24 commented on source/agora/consensus/data/Enrollment.d:
```diff
@@ -0,0 +1,43 @@
+import agora.common.crypto.Key;
+import agora.common.Deserializer;
+import agora.common.Hash;
+import agora.common.Serializer;
```
```suggestion
```

Geod24 commented on source/agora/consensus/data/Enrollment.d:
```diff
@@ -0,0 +1,43 @@
+import agora.common.crypto.ECC;
+import agora.common.crypto.Key;
+import agora.common.Deserializer;
+import agora.common.Hash;
```
```suggestion
```

Geod24 commented on source/agora/consensus/data/Enrollment.d:
```diff
@@ -0,0 +1,43 @@
+module agora.consensus.data.Enrollment;
+
+import agora.common.crypto.ECC;
+import agora.common.crypto.Key;
```
You can also remove those two and import `common.Types` instead

Geod24 commented on source/agora/common/EnrollmentManager.d:
```diff
@@ -0,0 +1,353 @@
+            return true;
+        }
+
+        return false;
```
This branch is not tested

Geod24 reviewed (commented):
Looks nice, just a few comments

linked0 commented:
@AndrejMitrovic Ping
